utilities/pytorch_model.py

```diff
@@ -200,32 +258,81 @@ def torch_model(self, X_train, X_test, y_train, y_test, params,
             model = model.cuda()
 
         # pos_weight is a scalar, the weight for the 1 class
-        criterion = nn.BCEWithLogitsLoss(pos_weight=pos_weight)
+        if classify:
+            criterion = nn.BCEWithLogitsLoss(pos_weight=pos_weight)
+        else:
+            criterion = nn.MSELoss()
         optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate)
-        scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(
-                        optimizer, patience=5)
 
         for epoch in range(num_epochs):
             running_loss = 0.0
             for i, (X_batch, y_batch) in enumerate(train_loader):
                 optimizer.zero_grad()
                 y_pred = model(X_batch)
                 loss = criterion(y_pred, y_batch)
+
                 # add l1 loss
                 l1_loss = sum(torch.norm(param, 1)
                                 for name, param in model.named_parameters()
                                 if 'bias' not in name)
                 loss += l1_penalty * l1_loss
+
+                # add L2 network penalty if applicable
+                if self.laplacian is not None:
+                    # get weights w/gradients
+                    w = [param for name, param in model.named_parameters()
+                               if 'bias' not in name][0]
+                    # filter for features that are in the network
+                    w = w[:, self.network_features]
+                    # calculate w^T @ L @ w
+                    network_loss = torch.mm(
+                        w.view(1, -1),
+                        torch.sparse.mm(self.laplacian, w.view(-1, 1)))
+                    loss += (network_penalty * network_loss).view(-1)[0]
+
                 running_loss += loss
                 loss.backward()
                 optimizer.step()
-            scheduler.step(running_loss)
+
+            if learning_curves:
+                # save train loss and test loss on whole dataset after each epoch
+                y_pred_train = model(X_tr)
+                y_pred_test = model(X_ts)
+                self.monitor_['train_loss'].append(float((
+                    criterion(y_pred_train, y_tr)
+                ).detach()))
+                self.monitor_['test_loss'].append(float((
+                    criterion(y_pred_test, y_ts)
+                ).detach()))
+                # also save l1 loss
+                self.monitor_['l1_loss'].append(float((
+                    sum(torch.norm(param, 1)
+                            for name, param in model.named_parameters()
+                            if 'bias' not in name)
+                ).detach()))
```

Comment:
Calling `.item()` here my fill the role of both the `float` cast and the call to `.detach()`. The parenthesis nesting is a little unclear to me though, so I'm not sure

Reply:
That seems to work!